### PR TITLE
Update JSONField for django 3.1+

### LINKED ISSUE #265: django 3.1: JSONField deprecation warning
I will update with a PR

`django.contrib.postgres.fields.JSONField` raises a deprecation warning in 3.1, and will be removed completely in 4.0.

See [Django 3.1 Release Notes](https://docs.djangoproject.com/en/3.1/releases/3.1/), section [JSONField for all supported database backends¶
](https://docs.djangoproject.com/en/3.1/releases/3.1/#jsonfield-for-all-supported-database-backends)

https://github.com/python-social-auth/social-app-django/blob/3e42f96fa68fda39c584bc2453faed0357c93720/social_django/fields.py#L12

#### CLI Warning

```
WARNINGS:
social_django.Partial.data: (fields.W904) django.contrib.postgres.fields.JSONField is deprecated. Support for it (except in historical migrations) will be removed in Django 4.0.
        HINT: Use django.db.models.JSONField instead.
social_django.UserSocialAuth.extra_data: (fields.W904) django.contrib.postgres.fields.JSONField is deprecated. Support for it (except in historical migrations) will be removed in Django 4.0.
        HINT: Use django.db.models.JSONField instead.
```

image

### PR DESCRIPTION
**Pull request by tony**
**Update 2020-09-06: I am not using this project anymore (due an outside reason, this project is fine). If there are edits needed that the maintainer can't add, feel free to fork [my branch](https://github.com/tony/social-app-django/tree/fix-jsonfield-import) / make your own branch and PR it. I apologize about this!**

Django 3.1 has a universal JSONField in `django.db.models`.
It raises a deprecation warning for JSONField imported via
`django.contrib.postgres`. Django 4.x removes it completely.

Preserve compatibility with older versions, import via compat module.
Hopefully this preserves migration compatibility.

Fixes #265

Also adds django 3.0 and 3.1 to CI - ignores `master` (which can fail sporadically)

# Changes

## CI

Add support for django 3.0 and 3.1 

## Settings

Django 3.1 introduces JSONField support for all backends and adds a deprecation
warning.

- Deprecating setting: `SOCIAL_AUTH_POSTGRES_JSONFIELD` (bool)

  Rename this to `SOCIAL_AUTH_JSONFIELD_ENABLED`. The setting will be deprecated in a future release.

- New setting: `SOCIAL_AUTH_JSONFIELD_ENABLED` (bool)

  Same behavior, setting name updated to match JSONField being supported by all systems.

- New setting: `SOCIAL_AUTH_JSONFIELD_CUSTOM` allows specifying an import string. This gives better control of the `JSONField` used by social-app-django.

  For django systems 3.1+, `django.db.models.JSONField` will be used by default. Older systems will fall back to `django.contrib.postgres.fields.JSONField`

  django 4.0': `django.contrib.postgres.fields.JSONField` [will still exist for import migrations](https://github.com/django/django/commit/7cb5712edc158396c9d4fbf1ecf17794d9a128b3#diff-f52b45a9d9ef14afdc1380fa97221caefb7ef3c9e71e67464d9719a91a87e64e)

  At any time, you can set to a custom JSONField of your own with, e.g.:

  `SOCIAL_AUTH_JSONFIELD_CUSTOM = 'django.contrib.postgres.fields.JSONField'`

### CODE REVIEW

**Comment by mrfredericks** [comment]
This field is useful for more db backends than just postgres.  Would it make more sense to have a distinct setting for using the new JSONField and then deprecate the POSTGRES_JSONFIELD setting?  I don't believe this change needs to be hidden from consumers as done here since this change may actually generate migrations, and they should be getting the deprecation warning mentioned. 

Thanks for putting this up 👍

**Comment by tony** [comment]
> This field is useful for more db backends than just postgres. Would it make more sense to have a distinct setting for using the new JSONField and then deprecate the POSTGRES_JSONFIELD setting? I don't believe this change needs to be hidden from consumers as done here since this change may actually generate migrations, and they should be getting the deprecation warning mentioned.

I've thought of that - but at the moment we still support older django versions and production environments using the old way, and I haven't found a way to do it while still not triggering a migration the user doesn't want.

I lean toward erroring on the side of safety due to that.

I think it's easier to remove the setting in subsequent versions where support for older django versions are dropped.

If we do want to remove `POSTGRES_JSONFIELD` setting eventually (maybe users are already relying on their own custom json fields, an angle I haven't thought of) - I think we should raise [`DeprecationWarning`](https://docs.python.org/3/library/exceptions.html#DeprecationWarning) to phase it out over a couple of releases.

**Comment by vschettino** [comment]
Could a migration from TextField to JSONField break something? Maybe the strategy `JSONField.check()` is better than the actual `POSTGRES_JSONFIELD` setting. For those using the old JSONField already, the change does not seem to add any incompability, as stated in the release notes:
-  https://docs.djangoproject.com/en/3.1/releases/3.1/#jsonfield-for-all-supported-database-backends

 If there's anything I can do to contribute I'll be glad to help!

**Comment by tony** [comment]
@vchettino @mrfredericks 

This PR adds forward compatibility, preserves backward compatibility and behavior.

I like the ideal of being able to declare `POSTGRES_JSONFIELD` since it gives existing projects control. This way the user can workaround a migration being created by setting it.

I'm not fully sure if `django.contrib.postgres.fields.JSONField` -> `django.db.models.JSONField` would trigger a migration. If it did, setting `POSTGRES_JSONFIELD` would be the workaround.

My intention with this PR is to have an update that risk-averse, and leave the conversation of larger changes (e.g. potentially deprecating `POSTGRES_JSONFIELD`) for larger release.

**Comment by vschettino** [comment]
@tony I agree this is the best strategy. Just to confirm, I checked and going from `django.contrib.postgres.fields.JSONField` to `django.db.models.JSONField` will generate a migration.

**Comment by tony** [comment]
@vschettino So...

I actually misunderstood `POSTGRES_JSONFIELD` - I thought for some magical reason it could be an import string to the field class. Eek.

Any thoughts? (@vschettino and anyone else interested)

This is django plugin - we shouldn't implicitly do something that'd create a migration (even if it's a noop).  We need to offer enough breathing room for the user to switch the field themselves.

Maybe we should actually continue to show the warning, but add a new setting to point the customize where the `extra_data` json field points to:

```
# below: Default for django < 3.whatever version, shows warning
SOCIAL_AUTH_JSONFIELD_CUSTOM = 'django.contrib.postgres.fields.JSONField'

# User can point to this, which eliminates warning and creates a new migratin
SOCIAL_AUTH_JSONFIELD_CUSTOM = 'django.db.models.JSONField'
```

Also one more thing:

We use the term `POSTGRES_JSONFIELD`, but jsonfield is (to be) universal. What we really probably want is something like `JSONFIELD_ENABLED` true/false.

**Comment by mrfredericks** [comment]
Yeah, the setting name might be a bit of a misnomer.  Unless there's actually another use case out there for setting a custom model to use, I think that a new true/false setting for the new JSONField is the simplest (for both us, and the consumer), and accomplishes all our goals.  After upgrading Django to 3.1 people can set that field at their leisure and run their normal migration process.

**Comment by tony** [comment]
I have issues coming up with enough time to get back to this.

I want to add a Setting variable to make the field class overridable so the user can choose when they want to use the new model.  That would make it so there's the best of both worlds: backward compatibility agnostic of django version/json field settings, while being able to use the latest django idioms on new installations

**Comment by vschettino** [comment]
Agree, having a new setting is better in backwards compability and gives people time/control to use the new field. I also agree it should be a boolean setting to keep it harmonic with the existing configuration. Maybe  It is a good opportunity to enhance the docs about the new setting and the old `POSTGRES_JSONFIELD` to avoid further confusion. Should we also deprecate the old setting? @tony tell me if there's anything I can do to help!

**Comment by tony** [comment]
@vschettino if you'd like to test the newest changes:

- Added `SOCIAL_AUTH_JSONFIELD_ENABLED`
- Deprecation warning for `SOCIAL_AUTH_POSTGRES_JSONFIELD`
- Added optional `SOCIAL_AUTH_JSONFIELD_CUSTOM`

Especially I'd like help testing out  `SOCIAL_AUTH_JSONFIELD_CUSTOM` to see if it works correctly with migrations, if possible

**Comment by vschettino** [comment]
Hey @tony, it works fine here! Thanks!

**Comment by tony** [comment]
@omab Any general feedback on this approach? (before I invest too much time in it)

(I don't use python-social-auth anymore on the project this was for, unfortunately)

**Comment by bellini666** [comment]
@tony when are those changes going to be released?

**Comment by tony** [comment]
@bellini666 In this case, I'm not a maintainer.  It's up to them

I pinged @omab here: https://github.com/python-social-auth/social-app-django/pull/266#issuecomment-694986079 but I won't be able to keep this rebased forever

I don't have indication if the general approach is okay or not. Someone may end up writing their own PR and ignoring this

In case I get busy: If it comes to if feel free to fork this PR and create a new one

**Comment by relwell** [comment]
Would be great to get this merged in. The deprecation warnings are a little bothersome!

**Comment by tony** [comment]
@omab Any thoughts on the direction I took with the PR?

Do you have a different idea for an approach?

Anything more needed from me? I'm not using this package anymore so I don't have as much ability to test it.

**Comment by tony** [comment]
@omab Hey there! Please read this. There is a warning message in Django 3.x for `JSONField`'s becoming standard across DBs. We need a migration path that satisfies older and newer django versions and deprecates old usage styles.

Any opinions on this approach? Anything else I can do? 

Let me know!

**Review by dulmandakh — CHANGES_REQUESTED** [review]
Could you please to master and see if tests are passing. Thanks

**Comment by tony** [comment]
@dulmandakh Rebased

For clarify, I will refer to settings name with the settings prefix included, e.g `POSTGRES_JSONFIELD` -> `SOCIAL_AUTH_POSTGRES_JSONFIELD`

## Recap

What's happening: There's a new [generic JSONField in django 3](https://docs.djangoproject.com/en/3.1/releases/3.1/#jsonfield-for-all-supported-database-backends), projects need to move over to it and social-app-django's hard-coded JSONField raises a warning.

###  Issues users are having:
- Users are moving to the new JSONField to 1.) Make the warning go away 2.) add JSONField for other DB's like mysql
- User projects need to update to the new JSONField in their own time

  We need to have release a version of social-app-django where projects can update their field

### Issues in this project
- social-app-django doesn't allow setting the JSONField

  Added `SOCIAL_AUTH_JSONFIELD_CUSTOM` to specify the field via import string

  This way the user can stay on the old field or go to the new one without triggering migrations/warnings.
- social-app-django's setting name references POSTGRES when json fields aren't postgres specific anymore

  We need to phase-out `SOCIAL_AUTH_POSTGRES_JSONFIELD`, to do so carefully:

  1. Add `SOCIAL_AUTH_JSONFIELD_ENABLED`
  2. Add deprecation warning for `SOCIAL_AUTH_POSTGRES_JSONFIELD` being removed and to renamed the setting to `SOCIAL_AUTH_JSONFIELD_ENABLED `

### After this

If this is merged, we can publish this in a feature release (e.g. 5.0)

Then, in 6.0 we can remove `SOCIAL_AUTH_POSTGRES_JSONFIELD` and any lingering references to POSTGRES.

File: social_django/fields.py
Comment (nijel) [inline social_django/fields.py:33]: This will switch all existing django.contrib.postgres.fields.JSONField users to django.db.models.JSONField once they are on Django 3.1. Is this intended behaviour? Won't this cause some problems?

**Review by nijel — COMMENTED** [review]
Thanks, the migration towards standard JSONField is certainly a good thing to do. 

The migration can't be done by just changing the existing field type, the database needs to be migrated as well. This is typically done by constructing temporary real JSONField, inserting old content there, removing the old one and renaming the temporary one. The tricky part is how to do this properly in a library while keeping interchangeable type for JSON field.

File: social_django/fields.py
Comment (tony) [reply social_django/fields.py:33]: That would raise an indication of some sort. I don't think there's an alternative / safe default in this instance, since we don't know the JSON Field the project is using.

They will be able to customize `JSONFIELD_CUSTOM`

Comment (nijel) [reply social_django/fields.py:33]: Sorry, I was not clearly describing the problematic case. 

The problem is with `POSTGRES_JSONFIELD` enabled. Before this change it was using `django.contrib.postgres.fields.JSONField`, but after this patch it changes to `django.db.models.JSONField` on Django 3.1+.

**Comment by tony** [comment]
@nijel

Thank you

> The migration can't be done by just changing the existing field type, the database needs to be migrated as well. 

Is this for switching from `django.contrib.postgres.fields.JSONField` or `models.JSONField` or from a `models.TextField` to json?

If it's textfield -> jsonfield, they have `JSONFIELD_ENABLED` to `False`. 

Secondly, I'm not sure if it's in the realm of responsibility for this project to handle something specific as a field migration for the user. 

Regardless of answer to the above question, Is there any prior art you know of where a django library provides that level of compatibility / a migration path such as that?

I think it's really up to making the model innards customizable to the user.   I'm not convinced there's a one-size-fits-all solution. For existing projects, this PR gives is the ability to opt-in/opt-out of the JSONField and customize the JSONField being pointed to. That lets them control how stuff is or isn't migrated over.

I believe system checks would catch the inconsistency and the user would be able to use `JSONFIELD_ENABLED` and `JSONFIELD_CUSTOM` to have social-app-django match their own project's jsonfield (or lack thereof)

**Comment by tony** [comment]
(^ I made various edits above)

**Comment by nijel** [comment]
I don't know what is needed for switching from `django.contrib.postgres.fields.JSONField` to `models.JSONField`. What I've described is for migration from `models.TextField` (and that is what most people will have because that is existing default).

The problem with leaving this up to the user is that it will easily break in more complex situations, for example in applications using PSA internally. I'm author of Weblate which relies on PSA and I would not do such migration because of this and PSA would stick there on `TextField` forever just to avoid this ;-).

Problems I see:

* The switchable models are tricky in migrations, it expects users to switch the configuration at the right time (we've done that while changing User model and it has proven to be problematic).
* Migrating PSA database from outside code seems ugly and easy to make it wrong. Especially when every user is expected to reinvent this. At least there needs to be some documentation on how to do that in a way that is less likely to break.
* The outside migration would depend on current PSA migrations, but any new PSA migration should  to be executed after it. This can be achieved using `run_before` on the outside migration, but that can be added only once there is new migration in PSA.

My preferred approach would be to PSA migrate to `models.JSONField` without making it configurable. The problem with this is that it probably can't be done now as it would make PSA require Django 3.1 or newer.

**Comment by tony** [comment]
@nijel 

> My preferred approach would be to PSA migrate to models.JSONField without making it configurable. The problem with this is that it probably can't be done now as it would make PSA require Django 3.1 or newer.

Does that entail removing `models.TextField` as a default and just making it `models.JSONField` for everything?

1. If so, what do you propose with me making a follow up PR that is django 3.1 or newer (including updating CI)
2. If above second PR proposal is okay, in regards to:

   > I'm author of Weblate which relies on PSA and I would not do such migration because of this and PSA would stick there on TextField forever just to avoid this ;-).

  What about also added `JSONFIELD_ENABLED` as a way to turn the field on / off being in the PR?  If so, what should be the default?

  If not, what would be the default behavior of when `models.TextField` is used?
3. If the second PR idea is okay: There could be a branch created, before the django 3.1 cutoff point, to maintain older django versions (and potentially release fixes to pypi for it)

What are your thoughts on this?

**Comment by nijel** [comment]
I don't feel like I should make this decision myself here (being PSA maintainer for less than day now). I think @omab should share his opinion as well.

Thinking about this again, I don't think it's a bad idea to introduce such configuration (given it doesn't break existing setups, see my comments on `POSTGRES_JSONFIELD` enabled). This would be a backwards compatible change allowing users to opt-in for a native JSON field. The long-term plan (let's say for PSA 6.0 to stick with versions you've outlined above) should then be to remove all these settings and fully migrate to native JSON field.

File: social_django/fields.py
Comment (tony) [reply social_django/fields.py]: I fixed two issues:
1. `django.contrib.postgres.fields.JSONField` is now the default option, unless an `ImportError` is raised. So there's no behavioral changes if the user upgrades. They can opt-in to the new field by setting `SOCIAL_AUTH_JSONFIELD_CUSTOM`.
2. If `SOCIAL_AUTH_JSONFIELD_CUSTOM` is enabled, that'll be the field used (before, it would override the custom field because no `else` was there)

**Comment by tony** [comment]
@nijel @omab  I made updated to assure that there is no behavioral changes for users upgrading to this PR

It'll still use the old `django.contrib.postgres.fields.JSONField` until it raises an `ImportError`. 

Users can control the json field via `JSONFIELD_CUSTOM`, so for 3.1+, users will be able to set:

`SOCIAL_AUTH_JSONFIELD_CUSTOM = 'django.db.models.JSONField'`

**Review by nijel — APPROVED** [review]
Thanks, looks good to me now.

I'm not sure about the long text in the changelog. The more detailed settings documentation should certainly go into https://github.com/python-social-auth/social-docs/blob/master/docs/configuration/settings.rst

File: social_django/fields.py
Comment (vytisb) [reply social_django/fields.py]: I think, this should be reverted to the original order. That is, first try to import from `django.db.models`, and if that fails then import from `django.contrib.postgres.fields`.

The `from django.contrib.postgres.fields import JSONField` import will never fail. It has to work to support historical migrations. The field is "removed" by having Django system checks fail when used in an actual model. This can be seen on [django master](https://github.com/django/django/blob/8fdb5a656a35dc9beef4cf2b79a782068263c7ec/django/contrib/postgres/fields/jsonb.py), which currently is the future Django 4.0.

Secondly, on Django 3.1+ the `django.contrib.postgres.fields.JSONField` is a subclass of `django.db.models.JSONField` adding nothing else than a system check warning. So everyone who previously used `django.contrib.postgres.fields.JSONField`, after upgrading to Django 3.1 immediately start using `django.db.models.JSONField`, and they only need to change the import location to avoid the warning.

Comment (nijel) [reply social_django/fields.py]: Ah, I didn't look into the Django code, but silently switching from `django.contrib.postgres.fields.JSONField` to `django.db.models.JSONField` seemed like an unwanted behaviour. If Django would do that anyway, then it's probably safe...

Comment (tony) [reply social_django/fields.py]: @vytisb @nijel I will switch this around (and update the PR README, Changelog, etc) accordingly.

And still - in either case - if there's an edgecase, the user can always set `SOCIAL_AUTH_JSONFIELD_CUSTOM` to override.

So we'll have both 1. A sensible, intuitive default that matches Django's behavior and 2. An ability for the user to handle edgecases without forking

**Comment by tony** [comment]
> I'm not sure about the long text in the changelog. The more detailed settings documentation should certainly go into https://github.com/python-social-auth/social-docs/blob/master/docs/configuration/settings.rst

I agree with adding the settings to `settings.rst`. I'm not sure I understand "long text" thing, since action needs to be taken with the variables based on the changes that happen when they upgrade the package. Here's an example from Django:

https://docs.djangoproject.com/en/3.1/releases/3.1/#jsonfield-for-all-supported-database-backends

File: social_django/fields.py
Comment (tony) [reply social_django/fields.py]: @vtisb @nijel I updated this to default to `django.db.models.JSONField` if it exists, and also rebased.

**Comment by tony** [comment]
@nijel @omab 

Before I go editing `settings.rst` (cloning the repo and making a PR for about the settings), is there a general thumbs up / thumbs down on this approach? 

It's a huge effort on my part to maintain this PR, I've probably put 10+ hours total of time into this over months (having to keep it rebased, having my CI fixes overwritten).  So I'd appreciate an indication if this is the way it's going.

**Review by nijel — APPROVED** [review]
I'm definitely okay with moving towards standard Django JSON field. I'd prefer to drop the configuration for this completely in the future, but that's not possible to do it right now.

File: social_django/fields.py
Comment (vytisb) [reply social_django/fields.py]: @tony I don't see the change. Maybe it got lost during rebase? Could you check again? Thanks.

Comment (tony) [reply social_django/fields.py]: @vparitskiy 

Good eye! How is it now?

**Comment by tony** [comment]
@nijel Thank you for letting me know!

So who would we need a sign-off from? @omab?  Anyone else?

File: social_django/fields.py
Comment (vytisb) [reply social_django/fields.py]: @tony Looks good. Thanks!

**Comment by nijel** [comment]
I don't think there is anybody besides @omab and me right now. He is usually slow to respond, so unless he speaks up, my review is probably as much as you can get. Many people expressed interest in https://github.com/python-social-auth/social-core/issues/445, but it doesn't seem that anybody of them is actually actively reviewing pull requests.

**Comment by omab** [comment]
@tony, I liked the approach on this one, at some point int he future I would like to migrate completely that configurable field to a JSON type, once we drop some oldish versions we can consider that and implement a built-in migration to handle the transition, meanwhile this looked great.

I've made a minor change in the changelog in order to provide a shorter entry and link to the oficial documentation where I've added a section for it.

**Comment by tony** [comment]
@omab This sounds like a plan. 

As for documentation, it looks like our bases are covered as well via https://github.com/python-social-auth/social-docs/pull/78 (Thank you for that!)

**Comment by selected-pixel-jameson** [comment]
Has this been released? I have `social-auth-app-django==4.0.0` installed but I'm still getting a warning.

`social_django.Partial.data: (fields.W904) django.contrib.postgres.fields.JSONField is deprecated. Support for it (except in historical migrations) will be removed in Django 4.0.`

**Comment by agconti** [comment]
Hi guys, thanks for all your work on this! How should existing users migrate a text field to a JSON field?

**Comment by agconti** [comment]
This solution also solves: https://github.com/python-social-auth/social-app-django/issues/209

**Comment by nijel** [comment]
Followup to this is dropping the customizability in https://github.com/python-social-auth/social-app-django/pull/465. All users will then migrate to native JSONField.

**Pull request merged**
